cvat/apps/engine/tests/utils.py

```diff
@@ -455,3 +459,80 @@ def filter_dict(
     d: dict[str, Any], *, keep: Sequence[str] = None, drop: Sequence[str] = None
 ) -> dict[str, Any]:
     return {k: v for k, v in d.items() if (not keep or k in keep) and (not drop or k not in drop)}
+
+
+def filter_object(
+    obj: list | dict, *, keep: Sequence[str] = None, drop: Sequence[str] = None
+) -> list | dict:
+    if isinstance(obj, dict):
+        obj = filter_dict(obj, keep=keep, drop=drop)
+        for k, v in obj.items():
+            obj[k] = filter_object(v, keep=keep, drop=drop)
+        return obj
+    elif isinstance(obj, list):
+        for i, val in enumerate(obj):
+            obj[i] = filter_object(val, keep=keep, drop=drop)
+        obj = list(filter(lambda x: x != dict(), obj))
+        return obj
+    return obj
+
+
+def freeze_object(obj, ignore_keys=None) -> frozenset:
+    if isinstance(obj, dict):
+        obj = filter_dict(obj, drop=ignore_keys)
+        for k, v in obj.items():
+            obj[k] = freeze_object(v, ignore_keys)
+        return frozenset(obj.items())
+    elif isinstance(obj, list):
+        for i, val in enumerate(obj):
+            obj[i] = freeze_object(val, ignore_keys)
+        return tuple(obj)
+    return obj
+
+
+def check_optional_fields(self: TestCase,
+                        obj: dict[str, int | float | str],
+                        optional_values: dict[str, int | float | str],
+                        # expected_values: dict[str, int | float | str]
+                        ) -> None | NoReturn:
+    for k in optional_values.keys():
+        # if obj.get(k) and not optional_values[k] == obj[k]:
+        self.assertEqual(obj.get(k, optional_values[k]), optional_values[k],
+                         f"key '{k}' has incorrect expected optional value {obj.get(k)} instead of {optional_values[k]}") # coalesce
+
+
+def compare_objects(self: TestCase, obj1, obj2, ignore_keys, fp_tolerance=0.001, current_key=None, order=True):
+    key_info = "{}: ".format(current_key) if current_key else ""
+    error_msg = "{}{} != {}"
+
+    def is_annotation_type(k):
+        return any(k.startswith(_type) for _type in [AnnotationType.SHAPE, AnnotationType.TAG])
+        # tracks are shapes, possibly nested
+
+    if isinstance(obj1, dict):
+        self.assertTrue(isinstance(obj2, dict), error_msg.format(key_info, obj1, obj2))
+        for k in (obj1.keys() - ignore_keys):
+            v1 = obj1[k]
+            v2 = obj2[k]
+            if k == "attributes":
+                key = lambda a: (a.get("spec_id") or a["id"])
+                v1.sort(key=key)
+                v2.sort(key=key)
+            elif not order and is_annotation_type(k):
+                v1 = frozenset(freeze_object(v1, ignore_keys))
+                v2 = frozenset(freeze_object(v2, ignore_keys))
+            compare_objects(self, v1, v2, ignore_keys, current_key=k)
```

Comment:
```suggestion
            compare_objects(self, v1, v2, ignore_keys, current_key=k, fp_tolerance=fp_tolerance, order=order)
```
And in the call below as well.